docs/adr/0008-use-tracing.md

```diff
@@ -0,0 +1,165 @@
+fails. Further, both our mobile apps and Firefox Desktop have techniques for having this log
+output be sent across the FFI so it can be captured in whatever "native" logging system is
+in place (eg, logcat on Android, custom file logging on iOS, Log.sys.mjs/Console.sys.mjs logging on Desktop)
+
+application-services also has an [error support crate](https://github.com/mozilla/application-services/tree/main/components/support/error),
+designed explicitly for error reporting for the applications.
+Android and iOS both hook into this to report errors to Sentry, whereas Desktop has not yet implemented this functionality.
+While the error support crate has capabilities beyond just moving the error across the FFI, this capability is the focus of this document,
+and any changes proposed to this crate by this document are limited to just this.
+
+### Problems with the current approaches.
+
+#### Problems using `log`
+
+The main problem with the `log` module is a design choice made by that crate,
+which is the concept of a [global "max level"](https://docs.rs/cli-log/latest/cli_log/fn.set_max_level.html)
+
+In Gecko, simply setting this max level to, say, `Trace` can cause performance regressions in other
+crates - eg, [see this desktop bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1874215) for one example.
+
+In practice, this means that we are unable to selectively get `Debug` or `Trace` logs for individual components,
+because the max level for the logger is, effectively, fixed at `Info`.
+
+### Problems with the error reporter
+
+The error reporter actually works OK for mobile, but is not implemented at all for Desktop.
+However, the existing implementation is quite ad-hoc and not very "rich" in terms of data which
+can be sent to the error reporter.
+
+So while there's actually no real problem here, there is an opportunity to better align the
+logging and error-reporting requirements into a single facility before introducing this capability
+to desktop.
+
+## The Rust tracing crate.
+
+An alternative to the `log` crate is the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate,
+which comes from the tokio project.
+This document is not going to rehash the excellent existing documentation, but some key features
+include:
+
+* For simple logging it has a very similar API to the `log` crate. Almost all existing use of `log` in non
+  test code can be upgrade to tracing simply by replacing `log::trace!` with `tracing::trace!` etc.
+* Richer logging semantics mean that it's quite easy and efficient to attach additional data to each log entry.
+  ie, more than simple strings can be logged.
+* A [`tracing-subscriber`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/) mechanism that allows
+  efficient filtering etc of log entries without enforcing a global `max_level`.
+  Each subscriber is able to efficiently filter log events without impacting other crates.
```

Comment:
How would this work?  Would it prevent the http3 formatting code from running as described in https://bugzilla.mozilla.org/show_bug.cgi?id=1874215?  I could imagine tracing only executing code if the subscriber visits the field, but I'm not sure if that's how it works.  If it does do that, then I think this should be called out even more, since it gives us the best of both worlds -- log records can be runtime configurable, with negligible performance overhead if they are filtered out.

Reply:
IIUC, the http3 crate made an explicit decision to check the max log level before deciding to do extra, expensive work. That seems like a bad idea in general, but isn't it correct to say it would be impossible to make that choice in this world?

Reply:
Ahh, so we could expose the log levels subscribed to by target, then http3 could use that to decide if they want to do the extra work.  That seems like an improvement over the current state of affairs for sure.

Reply:
I think that in practice, that kind of logging really isn't a great use-case for tracing in the first place. that http3 code is probably more than happy to just keep the status quo even if parts of it moved to logging.

IOW, in the general case I don't think `log` and `tracing` are mutually exclusive - maybe I should work out how to mention that here somewhere?

Reply:
> the http3 crate made an explicit decision to check the max log level before deciding to do extra, expensive work

Dogear used to do something like that: it checked the maximum log level before stringifying and printing the local, remote, and merged trees, so that it didn't do a bunch of work to build a huge string that might be immediately discarded if trace logging wasn't on.

We fixed that later by having trees implement `Display`—so they'd only be stringified in the actual `format!()` call that build the log message—but I'm not sure that what `http3` does now, or what Dogear used to do, is uncommon.